tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -0,0 +1,231 @@
+/*
+ * Parts derived from tests/kernel/fatal/src/main.c, which has the
+ * following copyright and license:
+ *
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <ztest.h>
+#include <kernel_structs.h>
+#include <string.h>
+#include <stdlib.h>
+
+#define INFO(fmt, ...) printk(fmt, ##__VA_ARGS__)
+
+/* ARM is a special case, in that k_thread_abort() does indeed return
+ * instead of calling _Swap() directly. The PendSV exception is queued
+ * and immediately fires upon completing the exception path; the faulting
+ * thread is never run again.
+ */
+#ifndef CONFIG_ARM
+FUNC_NORETURN
+#endif
+void _SysFatalErrorHandler(unsigned int reason, const NANO_ESF *pEsf)
+{
+	INFO("Caught system error -- reason %d\n", reason);
+	ztest_test_pass();
+#ifndef CONFIG_ARM
+	CODE_UNREACHABLE;
+#endif
+}
+
+static void is_usermode(void)
+{
+	/* Confirm that we are in fact running in user mode. */
+	zassert_true(_is_user_context(), "thread left in kernel mode\n");
+}
+
+static void priv_insn(void)
+{
+	/* Try to invoke what should require a privileged instruction. */
+	unsigned int key = irq_lock();
+
+	irq_unlock(key); /* just in case we succeeded */
+	zassert_unreachable("irq_lock did not fault\n");
+}
+
+static void write_control(void)
+{
+	/* Try to write to a control register. */
+#if defined(CONFIG_X86)
+	__asm__ volatile (
+		"mov %cr0, %eax;\n\t"
+		"and $0xfffeffff, %eax;\n\t"
+		"mov %eax, %cr0;\n\t"
+		);
+#elif defined(CONFIG_ARM)
+	__asm__ volatile (
+		"mrs r2, CONTROL;\n\t"
+		"bic r2, #1;\n\t"
+		"msr CONTROL, r2;\n\t"
+		);
+#endif
```

Comment:
Might be a good idea to add an #else block here with an #error directive, just so that, if another architecture ends up adding supervisor/user separation, this test won't fail silently.

Reply:
Done